pinot-core/src/main/java/org/apache/pinot/core/query/pruner/ValueBasedSegmentPruner.java

```diff
@@ -0,0 +1,202 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.query.pruner;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import org.apache.pinot.common.request.context.ExpressionContext;
+import org.apache.pinot.common.request.context.FilterContext;
+import org.apache.pinot.common.request.context.predicate.EqPredicate;
+import org.apache.pinot.common.request.context.predicate.InPredicate;
+import org.apache.pinot.common.request.context.predicate.Predicate;
+import org.apache.pinot.core.query.request.context.QueryContext;
+import org.apache.pinot.segment.local.segment.index.readers.bloom.GuavaBloomFilterReaderUtils;
+import org.apache.pinot.segment.spi.IndexSegment;
+import org.apache.pinot.segment.spi.datasource.DataSource;
+import org.apache.pinot.segment.spi.index.reader.BloomFilterReader;
+import org.apache.pinot.spi.data.FieldSpec.DataType;
+import org.apache.pinot.spi.env.PinotConfiguration;
+import org.apache.pinot.spi.exception.BadQueryRequestException;
+import org.apache.pinot.spi.utils.CommonConstants.Server;
+
+
+/**
+ * The {@code ValueBasedSegmentPruner} prunes segments based on values inside the filter and segment metadata and data.
+ */
+@SuppressWarnings({"rawtypes", "unchecked", "RedundantIfStatement"})
+abstract public class ValueBasedSegmentPruner implements SegmentPruner {
+  public static final String IN_PREDICATE_THRESHOLD = "inpredicate.threshold";
+  protected int _inPredicateThreshold;
+
+  @Override
+  public void init(PinotConfiguration config) {
+    _inPredicateThreshold =
+        config.getProperty(IN_PREDICATE_THRESHOLD, Server.DEFAULT_VALUE_PRUNER_IN_PREDICATE_THRESHOLD);
+  }
+
+  @Override
+  public boolean isApplicableTo(QueryContext query) {
```

Comment:
Suggest removing this so that the actual implementation must override it.
- For `ColumnValueSegmentPruner`, we can prune segments when there are EQ, IN (with less than threshold values), RANGE, and there is no NOT
- For `BloomFilterSegmentPruner`, we can prune segments when there are EQ, IN (with less than threshold values) and there is no NOT
- For both the pruners, if there is OR and one of the child cannot be pruned, the whole query is not prunable